Member event models

allouis opened the pull request:
refs https://github.com/TryGhost/Ghost/issues/12602

This adds models for each of the member event tables, as well as a migration to backfill members_paid_subscription_events table

allouis commented:
Migration here: https://github.com/TryGhost/Ghost/pull/12640/commits/1121e109600cbc94fddb6f11fee73f05a55bc2f1

matthanley commented on core/server/models/member-payment-event.js:
```diff
@@ -0,0 +1,47 @@
+const errors = require('@tryghost/errors');
+const ghostBookshelf = require('./base');
+
+const MemberPaymentEvent = ghostBookshelf.Model.extend({
```
Doesn't this model use the `members_payment_events` table [here](https://github.com/TryGhost/Ghost/blob/main/core/server/data/schema/schema.js#L358)? Also `.select(knex.raw('SUM(volume) as gross_volume'))` should be `.select(knex.raw('SUM(amount) as gross_volume'))`

allouis replied:
THANKS

matthanley commented on core/server/data/migrations/versions/4.0/14-populate-members-paid-subscription-events-table.js:
```diff
@@ -0,0 +1,112 @@
+                events.push(subscriptionCreatedEvent);
+            }
+
+            if (subscription.status === 'canceled') {
```
I don't think this will work as expected for cancelled events: if all of my subscriptions are cancelled, this migration will give me a negative aggregated MRR because we don't have the corresponding `created` events to cancel out. I think MRR delta for backfilling cancelled events should be 0.

allouis replied:
Okay maybe it's a bit too implicit, but the `else` block above will handle the initial subscription event so we would have an event with `+X` MRR and then the `-X` MRR afterwards

matthanley replied:
I completely missed that - but won't we actually get two lots of `-X` MRR since the status is still cancelled?

allouis replied:
Yes you're totally right, I'll move the modifier out of the calculateMrrDelta function

daniellockyer commented:
@allouis I think this looks ok but can we add a simple logging line somewhere like "Filling in historical members paid subscription events", just to make it easier to track migrations 🙂

allouis commented:
@daniellockyer dun!!

allouis committed:
Added MemberStatusEvent model

refs https://github.com/TryGhost/Ghost/issues/12602

allouis committed:
Added MemberPaymentEvent model

refs https://github.com/TryGhost/Ghost/issues/12602

- Adds the aggregatePaymentVolume option to the `findAll` method,
allowing us to calculate volume over time

allouis committed:
Added MemberLoginEvent model

refs https://github.com/TryGhost/Ghost/issues/12602

allouis committed:
Added MemberEmailChangeEvent model

refs https://github.com/TryGhost/Ghost/issues/12602

allouis committed:
Added MemberPaidSubscriptionEvent model

refs https://github.com/TryGhost/Ghost/issues/12602

- Adds the aggregateMRRDeltas option to the `findAll` method,
allowing us to calculate MRR over time

allouis committed:
Populated members_paid_subscription_events table

refs https://github.com/TryGhost/Ghost/issues/12602

By backfilling the data, we allow exisiting sites to view their MRR over
time
